diff --git a/kcfile.cc b/kcfile.cc
index 1e43d41..6bfc17a 100644
--- a/kcfile.cc
+++ b/kcfile.cc
@@ -1520,7 +1520,7 @@ char* File::read_file(const std::string& path, int64_t* sp, int64_t limit) {
   ::LARGE_INTEGER sbuf;
   if (!::GetFileSizeEx(fh, &sbuf)) {
     ::CloseHandle(fh);
-    return false;
+    return NULL;
   }
   if (limit > (int64_t)sbuf.QuadPart) limit = sbuf.QuadPart;
   char* buf = new char[limit+1];
diff --git a/kcthread.cc b/kcthread.cc
index 7db3114..57f2177 100644
--- a/kcthread.cc
+++ b/kcthread.cc
@@ -1969,7 +1969,7 @@ TSDKey::TSDKey() : opq_(NULL) {
   _assert_(true);
   ::DWORD key = ::TlsAlloc();
   if (key == 0xFFFFFFFF) throw std::runtime_error("TlsAlloc");
-  opq_ = (void*)key;
+  opq_ = reinterpret_cast<void*>(static_cast<::DWORD_PTR>(key));
 #else
   _assert_(true);
   ::pthread_key_t* key = new ::pthread_key_t;
@@ -1988,7 +1988,7 @@ TSDKey::TSDKey(void (*dstr)(void*)) : opq_(NULL) {
   _assert_(true);
   ::DWORD key = ::TlsAlloc();
   if (key == 0xFFFFFFFF) throw std::runtime_error("TlsAlloc");
-  opq_ = (void*)key;
+  opq_ = reinterpret_cast<void*>(static_cast<::DWORD_PTR>(key));
 #else
   _assert_(true);
   ::pthread_key_t* key = new ::pthread_key_t;
@@ -2005,7 +2005,7 @@ TSDKey::TSDKey(void (*dstr)(void*)) : opq_(NULL) {
 TSDKey::~TSDKey() {
 #if defined(_SYS_MSVC_) || defined(_SYS_MINGW_)
   _assert_(true);
-  ::DWORD key = (::DWORD)opq_;
+  ::DWORD key = static_cast<::DWORD>(reinterpret_cast<::DWORD_PTR>(opq_));
   ::TlsFree(key);
 #else
   _assert_(true);
@@ -2022,7 +2022,7 @@ TSDKey::~TSDKey() {
 void TSDKey::set(void* ptr) {
 #if defined(_SYS_MSVC_) || defined(_SYS_MINGW_)
   _assert_(true);
-  ::DWORD key = (::DWORD)opq_;
+  ::DWORD key = static_cast<::DWORD>(reinterpret_cast<::DWORD_PTR>(opq_));
   if (!::TlsSetValue(key, ptr)) std::runtime_error("TlsSetValue");
 #else
   _assert_(true);
@@ -2039,7 +2039,7 @@ void TSDKey::set(void* ptr) {
 void* TSDKey::get() const {
 #if defined(_SYS_MSVC_) || defined(_SYS_MINGW_)
   _assert_(true);
-  ::DWORD key = (::DWORD)opq_;
+  ::DWORD key = static_cast<::DWORD>(reinterpret_cast<::DWORD_PTR>(opq_));
   return ::TlsGetValue(key);
 #else
   _assert_(true);
